src/dataflow/operators/unordered_input.rs

```diff
@@ -0,0 +1,121 @@
+
+use std::rc::Rc;
+use std::cell::RefCell;
+use std::default::Default;
+
+use progress::frontier::{MutableAntichain, Antichain};
+use progress::{Operate, Timestamp};
+use progress::nested::subgraph::Source;
+use progress::count_map::CountMap;
+use progress::timestamp::RootTimestamp;
+use progress::nested::product::Product;
+
+use timely_communication::Allocate;
+use {Data, Push};
+use dataflow::channels::Content;
+use dataflow::channels::pushers::{Tee, Counter as PushCounter};
+use dataflow::channels::pushers::buffer::{Buffer as PushBuffer, Session};
+
+use dataflow::operators::Capability;
+use dataflow::operators::capability::mint as mint_capability;
+
+use dataflow::{Stream, Scope};
+use dataflow::scopes::{Child, Root};
+
+/// Create a new `Stream` and `Handle` through which to supply input.
+pub trait UnorderedInput<G: Scope> {
+    /// Create a new `Stream` and `Handle` through which to supply input.
+    ///
+    /// The `new_input` method returns a pair `(Handle, Stream)` where the `Stream` can be used
+    /// immediately for timely dataflow construction, and the `Handle` is later used to introduce
+    /// data into the timely dataflow computation.
+    ///
+    /// The `Handle` also provides a means to indicate
+    /// to timely dataflow that the input has advanced beyond certain timestamps, allowing timely
+    /// to issue progress notifications.
+    /// ```
+    fn new_unordered_input<D:Data>(&mut self) -> ((UnorderedHandle<G, D>, Capability<G::Timestamp>), Stream<G, D>);
+}
+
+
+impl<G: Scope> UnorderedInput<G> for G {
+    fn new_unordered_input<D:Data>(&mut self) -> ((UnorderedHandle<G, D>, Capability<G::Timestamp>), Stream<G, D>) {
+
+        let (output, registrar) = Tee::<G::Timestamp, D>::new();
+        let produced = Rc::new(RefCell::new(CountMap::new()));
+        let (cap, helper) = UnorderedHandle::new(PushCounter::new(output, produced.clone()));
+        let copies = self.peers();
+
+        let index = self.add_operator(UnorderedOperator {
+            progress: helper.progress.clone(),
+            messages: produced.clone(),
+            copies:   copies,
+        });
+
+        return ((helper, cap), Stream::new(Source { index: index, port: 0 }, registrar, self.clone()));
+    }
+}
+
+struct UnorderedOperator<T:Timestamp> {
+    progress:   Rc<RefCell<CountMap<T>>>,           // times closed since last asked
+    messages:   Rc<RefCell<CountMap<T>>>,           // messages sent since last asked
+    copies:     usize,
+}
+
+impl<T:Timestamp> Operate<T> for UnorderedOperator<T> {
+    fn name(&self) -> String { "Input".to_owned() }
+    fn inputs(&self) -> usize { 0 }
+    fn outputs(&self) -> usize { 1 }
+
+    fn get_internal_summary(&mut self) -> (Vec<Vec<Antichain<<T as Timestamp>::Summary>>>,
+                                           Vec<CountMap<T>>) {
+        let mut map = CountMap::new();
+        map.update(&Default::default(), self.copies as i64 - 1);
```

Comment:
Yea, I wasn't super-careful with this. I was paying more attention to the API at this stage (e.g. it may be cool to auto-flush an epoch once its capability is dropped, which can't be done right now). Let me know if you think this chunk of code would be, in general, a good idea and I'll try and fix the "race" with your feedback.